src/nupic/ntypes/Array.hpp

```diff
@@ -195,92 +236,136 @@ class Array : public ArrayBase {
+  template <typename T> 
+  std::vector<T> asVector() const {
+    std::vector<T> v;
+    NTA_BasicType to_type = BasicType::getType<T>();
+    if (has_buffer()) {
+      v.resize(count_);
+      BasicType::convertArray(v.data(), to_type, 
+                              getBuffer(), getType(), getCount());
+    }
     return v;
   }
 
   /**
-   * from a vector; copies buffer into this object,
-   *   example:   array.fromVector(v);
+   * from a vector; copies dense buffer into this object, with no type
+   * conversion. Array type becomes the type of the vector elements. example:
+   * array.fromDenseVector(v);
+   * Note: a dense array is a normal sequence of data as opposed to a sparse array
+   *       which is a sequence of indexes of non-zero values.
    */
-  template <typename T> void fromVector(std::vector<T> &vect) {
+  template <typename T> 
+  void fromDenseVector(const std::vector<T> &vect) {
     type_ = BasicType::getType<T>();
     allocateBuffer(vect.size());
-    memcpy(getBuffer(), vect.data(), count_ * BasicType::getSize(type_));
+    if (has_buffer())
+      memcpy(getBuffer(), vect.data(), count_ * BasicType::getSize(type_));
   }
-
   /**
-   * Type conversion
-   * This will do a full copy, converting to target type,
-   * and return the new Array.
+   * from a vector; copies sparse buffer into this object, with no type
+   * conversion. Array type becomes SDR with its Flat Sparse array populated.
+   *   example:   array.fromSparseVector(sparse_v, dim);
+   * Note: a dense array is a normal sequence of data as opposed to a sparse array
+   *       which is a sequence of indexes of non-zero values.
+   *
+   * args: 
+   *  vect  - a vector containing indexes of non-zero values.
+   *  dim   - a vector which defines the size/shape of the original dense SDR buffer.
    */
+  void fromSparseVector(const std::vector<Int32> &vect, const Dimensions &dim) {
+    size_t s = 1;
+    for (UInt d : dim)
+      s *= d;
+    NTA_CHECK(dim.size() == 0 || s == 0) << "fromSparseVector(); no dimensions given.";
+    if (type_ != NTA_BasicType_SDR || getSDR()->dimensions != dim) {
+      // re-allocate the buffer if needed.
+      type_ = NTA_BasicType_SDR;
+      allocateBuffer(dim);
+    }
+    RefreshCache(); 
+    getSDR()->setFlatSparse(vect);
+  }
+  /**
+    * Type Conversion
+    * This populates our Array from a templeted vector.
+    * The buffer is copied and a coversion is provided if
+    * vector element types are not the same as the Array type.
+    * The vector is assumed to contain only dense format data.
+    */
+  template <typename T> void populate(const std::vector<T> &v) {
+    NTA_BasicType fromType = BasicType::getType<T>();
+    allocateBuffer(v.size());
+    if (getCount() > 0) {
+      if (getType() == NTA_BasicType_Sparse) {
+        UInt32 j = 0u;
+        UInt32 *Destptr = (UInt32 *)getBuffer();
+        for (UInt32 i = 0u; i < getCount(); i++) {
+          if (v[i])
+            Destptr[j++] = (UInt32)i;
```

Comment:
> otherwise this would crash, ie for T=string

I don't think it would compile for a string.  It must be one of the valid temples. See BasicType.cpp line 100.